src/hotspot/share/utilities/lineReader.cpp

```diff
@@ -0,0 +1,126 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "runtime/os.hpp"
+#include "utilities/lineReader.hpp"
+
+LineReader::LineReader() {
+  _file = nullptr;
+  _is_oom = false;
+  _buffer = nullptr;
+  _buffer_len = 0;
+}
+
+LineReader::LineReader(FILE* file) {
+  _is_oom = false;
+  _buffer = nullptr;
+  _buffer_len = 0;
+  init(file);
+}
+
+LineReader::~LineReader() {
+  if (_buffer != nullptr) {
+    os::free(_buffer);
+  }
+}
+
+void LineReader::init(FILE* file) {
+  assert(file != nullptr, "sanity");
+  _file = file;
+  _buffer_len = DEBUG_ONLY(16) NOT_DEBUG(4096); // debug build: start small to test expansion logic
+  _buffer = (char*)os::malloc(_buffer_len, mtClass);
+  if (_buffer == nullptr) {
+    _is_oom = true;
+  }
+}
+
+char* LineReader::read_line() {
+  STATIC_ASSERT(0 < MAX_LEN && MAX_LEN <= INT_MAX);
+  assert(_file != nullptr, "must be initialized");
+  if (_is_oom) {
+    return nullptr;
+  }
+
+  int line_len = 0; // the number of characters we have read so far (excluding the trailing \0)
+  while (true) {
+    assert(line_len < _buffer_len, "sanity");
+    int available_space = _buffer_len - line_len;
+    char* p = _buffer + line_len;
+    int new_len = 0;
+    if (fgets(p, available_space, _file) == nullptr) {
+      // _file is at EOF
+      if (line_len == 0) {
+        return nullptr; // EOF
+      } else {
+        // We have read something in previous loop iteration(s). Return that.
+        // The next call to read_line() will return nullptr to indicate EOF.
+        return _buffer;
+      }
+    }
+
+    // fgets() reads at most available_space chars, including the trailing \0, so
+    // strlen(p) must be smaller than INT_MAX, and can be safely cast to int.
+    assert(strlen(p) < INT_MAX, "sanity");
+    new_len = (int)strlen(p);
+
+    // _buffer_len will stop at MAX_LEN, so we will never be able to read more than
+    // MAX_LEN chars for a single input line.
+    assert(line_len >= 0 && new_len >= 0 && (line_len + new_len) >= 0, "no int overflow");
+
+    line_len += new_len; // We have read line_len chars so far.
+
+    assert(line_len < _buffer_len, "sanity");
+    assert(_buffer[line_len] == '\0', "sanity");
+
+    if (_buffer[line_len - 1] == '\n' || feof(_file)) {
+      // We have read an entire line, or reached EOF
+      return _buffer;
+    }
+
+    if (line_len == _buffer_len - 1) {
+      // The buffer is not big enough to hold the entire input line. Expand it.
+      if (_buffer_len == MAX_LEN) {
+        // Cannot expand anymore. Return the first MAX_LEN-1 bytes of input.
+        // The behavior is exactly the same as if we had called fgets() with a
+        // buffer whose size is MAX_LEN.
+        return _buffer;
+      }
+      int new_len = _buffer_len * 2;
```

Comment:
Again UB on the overflow check. MAX_LEN should be set so that doubling of the current size will always hit max_len so you can't overflow.

Reply:
Fixed